keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2300 @@
+unless there is no other maintenance in progress.
+- If there are foreign finalizers on the Evacuation, it should only remove its own instigator
+  finalizer (see [Drain](#drain)).
+- If the evacuator does not support a cancellation and it has set
+  `.status.evacuationCancellationPolicy` to `Forbid`, deletion of the Evacuation object will not be
+  attempted.
+
+Consequences for pods:
+1. Pods whose evacuators have not yet initiated evacuation will continue to run unchanged.
+2. Pods whose evacuators have initiated evacuation and support cancellation
+   (`.status.evacuationCancellationPolicy=Allow`) should cancel the evacuation and keep the pods
+   available.
+3. Pods whose evacuators have initiated evacuation and do not support cancellation
+   (`.status.evacuationCancellationPolicy=Forbid`) should continue the evacuation and eventually
+   terminate the pods.
+
+#### Drain
+
+When a `Drain` stage is detected on the NodeMaintenance object, Evacuation objects are created for
+selected pods ([Pod Selection](#pod-selection)).
+
+```yaml
+apiVersion: v1alpha1
+kind: Evacuation
+metadata:
+  finalizers:
+    evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io
+  name: f5823a89-e03f-4752-b013-445643b8c7a0-muffin-orders-6b59d9cb88-ks7wb
+  namespace: blue-deployment
+spec:
+  podRef:
+    name: muffin-orders-6b59d9cb88-ks7wb
+    uid:  f5823a89-e03f-4752-b013-445643b8c7a0
+  progressDeadlineSeconds: 1800
+
+```
+
+This is resolved to the following Evacuation object according to the pod on admission:
+
+```yaml
+
+apiVersion: v1alpha1
+kind: Evacuation
+metadata:
+  finalizers:
+    evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io
+  labels:
+    app: muffin-orders
+  name: f5823a89-e03f-4752-b013-445643b8c7a0-muffin-orders-6b59d9cb88-ks7wb
+  namespace: blue-deployment
+spec:
+  podRef:
+    name: muffin-orders-6b59d9cb88-ks7wb
+    uid:  f5823a89-e03f-4752-b013-445643b8c7a0
+  progressDeadlineSeconds: 1800
+  evacuators:
+    - evacuatorClass: deployment.apps.k8s.io
+      priority: 10000
+      role: controller
+```
+
+The node maintenance controller requests the removal of a pod from a node by the presence of the
+Evacuation. Setting `progressDeadlineSeconds` to  1800 (30m) should give potential evacuators
+enough time to recover from a disruption and continue with the graceful evacuation. If the
+evacuators are unable to evacuate the pod, or if there are no evacuators, the evacuation controller
+will attempt to evict these pods, until they are deleted.
+
+The only job of the node maintenance controller is to make sure that the Evacuation object exist
+and has the `evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io` finalizer. 
+
+#### Pod Selection
+
+The pods for evacuation would first be selected by node (`.spec.nodeSelector`). NodeMaintenance
+should eventually remove all the pods from each node. To do this in a graceful manner, the
```

Comment:
+ 💯 to allow definition of rules allowing to intentionally skip drain for a set of Pods, it is a request that popped up several time in Cluster API, mostly due to folks deploying stuff tolerating all taints or just the ones applied by drain.

Notably, it is also worth to notice that users don't always have the chance to change Pods, e.g. because they are managed by some operator, because the app is installed by something that keep the spec in sync etc etc.
In those cases, skip drain could provide a nice exit path for operators trying to unblock node deletion.

Reply:
> I'm not entirely sure if the order is configurable (see questions above),
it is not

> This means that e.g. monitoring agents (logging, metrics, ...) that are usually deployed via DaemonSets will be terminated before static pods. Which in turn means that we won't have visibility during the last stages of the drain. I think this would make it hard to detect / troubleshoot the last parts of the drain.

This is similar to today: DaemonSets are also terminated before static pods.

> I'm also wondering in general if it would make sense to allow to skip drain for certain Pods:

> +💯 to add support for rules allowing to intentionally skip drain for a set of Pods, it is a request that popped up several time in Cluster API, mostly due to folks deploying stuff tolerating all taints or just the ones applied by drain.

I think the presumption is to better tune the types of workloads and their priorities. The goal of the NodeMaintenance is to remove all pods from a node and most likely to take the node offline.

> Notably, it is also worth to notice that users don't always have the chance to change Pods, e.g. because they are managed by some operator, because the app is installed by something that keep the spec in sync etc etc.
In those cases, skip drain could provide a nice exit path for operators trying to unblock node deletion.

> What about Pods that are tolerating all taints. I think this is a valid use case as sometimes folks want to enforce that Pods are running on all Nodes independent of whatever taints might be on that Node (for vendors it's impossible to tell which custom taints folks might use).
> - I think in that scenario a Node drain will be just stuck indefinitely. We are able to evict/evacuate these Pods, but the kube-scheduler will just reschedule them on the same Node.
> - Because there can be different personas involved it's also not always possible for the cluster-admin to simply modify tolerations of "third-party" Pods

These are the problems that the NodeMaintenance+Evacuation API tries to solve. Operators and any other actor should be aware of NodeMaintenance and Evacuation objects and they should offer the exit path automatically. And not block the drain and wait for manual intervention.

Reply:
> This is similar to today: DaemonSets are also terminated before static pods.

Is this a part of the kubelet shutdown feature?